### ISSUE TITLE
ES6 syntax failing in 0.3.4

### ISSUE THREAD

**Issue by aweary** [lukeed/taskr#43]
When trying to test the quickstart guide against `0.3.4` I found that `fly` is now throwing an error when running or query tasks from a `Flyfile` with ES6 syntax.

![Fly failing on ES6](http://i.imgur.com/is0Hdpj.gif)

``` js
stack: 
  - ~/.dev/flytest/Flyfile.babel.js:1
  - (function (exports, require, module, __filename, __dirname) { export default f
  -                                                               ^^^^^^
  - SyntaxError: Unexpected token export
  -     at exports.runInThisContext (vm.js 73:16)
  -     at Module._compile (module.js 443:25)
  -     at Object.Module._extensions..js (module.js 478:10)
  -     at Module.load (module.js 355:32)
  -     at Function.Module._load (module.js 310:12)
  -     at Module.require (module.js 365:17)
  -     at require (module.js 384:17)
  -     at Object.list (~/.npm-packages/lib/node_modules/fly/dist/cli/list.js 28:14)
  -     at callee$0$0$ (~/.npm-packages/lib/node_modules/fly/dist/index.js 81:13)
  -     at tryCatch (~/.npm-packages/lib/node_modules/fly/node_modules/babel-runtime/regenerator/runtime.js 67:40)
message: 
  """
    ~/.dev/flytest/Flyfile.babel.js:1
    (function (exports, require, module, __filename, __dirname) { export default f
                                                                  ^^^^^^
    Unexpected token export
  """

```

This did not occur prior to `0.3.4` as far as I tested, and I am able to reproduce with a clean install of `fly` on my machine. This also does not occur when using ES5 syntax.

**Comment by ghost** [comment]
I tried to reproduce this with a clean install, but I couldn't. The currently listed deps seem OK too. I know this occurs when babel-core does not exist in your `node_modules`.

**Comment by aweary** [comment]
Looks like `babel-core` does indeed exist in the `node_modules` folder for fly, so I don't think thats it. Let me try it out in a vagrant instance and see if I can reproduce.

**Comment by ghost** [comment]
Appreciated.

**Comment by ghost** [comment]
@Aweary What's the status of this issue? I do realize by looking at your screencast that the file is correctly named `Flyfile.babel.js`, however I would like to say for the record, that if you were using `Flyfile.js` it would not work according to the docs:

> Flyfiles must include the extension of the language they are written in, .js for ES5, .babel.js for ES6/7, .coffee for CoffeeScript, etc.

**Comment by aweary** [comment]
@bucaran I wasn't able to reproduce this issue in a vagrant box running `precise64`, but the issue still occurs for me on my MBP. It might be related to some packages I have installed globally that also happen to be dependencies, though I updated them all and still experienced this issue.

One thing I also noticed is that `npm i fly` is not sufficient to get the CLI running, so the guide probably needs to mention the recommended fix, whether it be `npm run setup` or just using `npm i -g fly`

**Comment by MadcapJake** [comment]
`npm i -D fly` works for me but I use an npm script in my `package.json`:

``` js
"scripts": {
    "fly": "fly"
}
```

Then just `npm run fly` and it runs the local instance of fly. No other setup required. ES6/CoffeeScript/EarlGrey works right out of the box this way for me.

**Comment by Aweary** [comment]
That's definitely one way to do it, but if you tried to run `flybut the documentation (specifically looking at the quickstart guide) needs to specify, because the current instructions don't actually work, and I don't think adding and running `npm` scripts is assumed to be the default setup process. `npm install -g fly` works and then lets you call `fly` from the CL directly, but I'm not sure about the implications of a global install with this package.

Anyways,

**Comment by aweary** [comment]
That's definitely one way to do it, but really what I was saying is that the documentation (specifically looking at the quickstart guide) needs to specify, because the current instructions don't actually work. 

I don't think adding and running `npm` scripts is assumed to be the default setup process. `npm install -g fly` works and then lets you call `fly` from the CL directly, but I'm not sure about the implications of a global install with this package.

Anyways, I thought it might be my global install of `babel` (you can see the trace shows thats where babel was being invoked) but I removed my global install and it's still giving me this error. Here's the new trace:

``` js
stack: 
  - ~/.dev/_flyjs/Flyfile.babel.js:1
  - (function (exports, require, module, __filename, __dirname) { export default f
  -                                                               ^^^^^^
  - SyntaxError: Unexpected token export
  -     at exports.runInThisContext (vm.js 73:16)
  -     at Module._compile (module.js 443:25)
  -     at Object.Module._extensions..js (module.js 478:10)
  -     at Module.load (module.js 355:32)
  -     at Function.Module._load (module.js 310:12)
  -     at Module.require (module.js 365:17)
  -     at require (module.js 384:17)
  -     at Object.spawn$ (~/.dev/_flyjs/node_modules/fly/dist/cli/spawn.js 29:16)
  -     at tryCatch (~/.dev/_flyjs/node_modules/fly/node_modules/babel-runtime/regenerator/runtime.js 67:40)
  -     at GeneratorFunctionPrototype.invoke [as _invoke] (~/.dev/_flyjs/node_modules/fly/node_modules/babel-runtime/regenerator/runtime.js 315:22)
message: 
  """
    ~/.dev/_flyjs/Flyfile.babel.js:1
    (function (exports, require, module, __filename, __dirname) { export default f
                                                                  ^^^^^^
    Unexpected token export
  """
```

I'm using `v0.12.0` for node and running `OS X 10.10 (14A361c)` on a MacBook Pro (Retina, 13-inch, Late 2013), for reference.

**Comment by ghost** [comment]
@Aweary Thanks for following us up on this.

# tl;dr
1. If you did `npm i -g fly` you _should be_ OK, otherwise let's file this as a bug.
2. If you cloned Fly and then tried to run `fly -f examples/`, please see → [docs/README.md#hacking](https://github.com/flyjs/fly/blob/master/docs/README.md#hacking)
3. If you did `npm i fly` **make sure to**:
   - have `node_modules/.bin` in your `$PATH` _or_,
   - symlink `./bin/index.js` to `/usr/local/bin` (or your preferred binaries path).
   - To symlink, simply run `npm run setup` after you `npm i fly`

#### Long Version

I think the Quickstart guide needs to be revised ASAP. I'll work on that today :+1: 

To understand the implications of installing Fly globally, just look at the `dependencies` field in `package.json`. _You should be OK_. Now, I don't know what is your preferred workflow with node packages in general, but I don't install anything globally, so...

I just **append** → `node_modules/.bin` to my `$PATH` so that I can run whatever command a package exports as a binary (only works when I am inside a directory that has a `node_modules/` directory with the module).

> When your shell session starts, in your shell init configuration file, in my case fish, it looks kind of like this: `set PATH $PATH node_modules/.bin`

**Now**, regarding the issue, excuse me if I am mistaken, but are you (1) hacking on Fly or (2) using Fly as a build too?

If (1) that means you cloned the repo and are possibly trying to run the `examples/` or fiddle with the code. There is [documentation for that here](https://github.com/flyjs/fly/blob/master/docs/README.md#hacking).

If (2) that means you simply `npm i fly` inside a directory and are running `fly` from there. If you installed `fly` globally you should be able to run `fly`, but if you didn't then you **can't**, unless you have `node_modules/.bin` in your `$PATH`, which is always a good idea anyway, or symlink `./bin/index.js` to a path included in `$PATH`.

If you don't want to `npm -g fly` or add `node_modules/.bin` to your `$PATH` then you need to symlink `./bin/index.js` to somewhere like `/usr/local/bin`. Fortunately there is a little script for that in `package.json`. Just run `npm run setup` after you `npm install fly` in any directory.

I will make sure to add a more complex `Flyfile.babel.js` that does the work of all these scripts in the next milestone, but I think the problem has to do more with how you were using Fly, that is, the (1) or (2) points from above.

Does this resolve your issue?

Regards

**Comment by ghost** [comment]
> Anyways, I thought it might be my global install of babel (you can see the trace shows thats where babel was being invoked) but I removed my global install and it's still giving me this error.

Interesting, I will test things on my end with a global `babel` install and see if the problem persists.

When we compile Flyfiles on the _fly_, we need to pass certain flags to babel to opt in all the features we provide in Fly, so your issue could be that. 

Just out of curiosity, try `npm uninstall babel`, and run over your experiments again. If you are able to run Fly without errors, then we could pinpoint the problem is in fact because of the global babel installation.

**Comment by ghost** [comment]
@Aweary OK, upon further review of the Quickstart tutorial and re-reading your post I think the problem is simply that you don't have `node_modules/.bin` in your `$PATH`. So, you have to either add it and run `fly` inside the directory into which you install Fly or install Fly globally.

@Aweary @MadcapJake 
I have updated the Quickstart guide with the following notice:

> If you can't run Fly is because you `node_modules/.bin` is not in your $PATH. You can either add it in POSIX shell `PATH=$PATH:node_modules/.bin` or in fish `set PATH $PATH node_modules/.bin` or install Fly globally `npm i -g fly`.

**Comment by aweary** [comment]
To clarify, this issue does occur when `fly` is installed globally as well.

**Comment by ghost** [comment]
Even after you uninstall a globally installed `babel`?

Also, can you show us the output of `ls /usr/local/lib/node_modules`?

**Comment by aweary** [comment]
Yupp. I even created a new admin user on my computer, used `nvm` to install and use node `0.12.7`, so the global package would be isolated, and installed `fly` globally and the same exact thing occurs. 

`/usr/local/lib/node_modules` just contains `npm`, though my global packages are installed in `~./node_modues` if thats what you're looking for.

**Comment by ghost** [comment]
@Aweary I tried installing Fly globally, but everything ran fine. It's difficult to debug this if I am unable to reproduce the bug :disappointed: 

Do you also get an error if you use a regular ES5 `Flyfile.js` file?

**Comment by aweary** [comment]
ES5 syntax does not throw this error. Are you using the same platform? I'm running `OS X 10.10 (14A361c)` and I've reproduced using node `0.12.7` and `0.11`.

Have you tried creating a new user and testing to 100% rule out the possibility of `fly` using previously installed dependencies somewhere? I've gone as far as completely reinstalling `node` and `npm` and removing all global packages. Issue still occurs. I'm worried it might be a `fly` or `babel` issue that only occurs on certain platforms.

**Comment by ghost** [comment]
I am using the same platform. This is clearly something specific to your system because other folks have not been able to reproduce your bug. You would help us tremendously if you found a way to reproduce the bug.

#### Current Fly dependencies

``` json
  "dependencies": {
    "babel-core": "^5.7.4",
    "babel-runtime": "^5.6.18",
    "clor": "^1.0.0",
    "co": "^4.5.4",
    "fly-util": "^3.4.3",
    "glob": "^5.0.13",
    "mkdirp": "^0.5.1",
    "mz": "^2.0.0",
    "parsec": "^0.1.0",
    "prettyjson": "^1.1.2",
    "rimraf": "^2.4.0",
    "yieldables": "^1.2.0"
  }
```

**Comment by MadcapJake** [comment]
> /usr/local/lib/node_modules just contains npm, though my global packages are installed in ~./node_modues if thats what you're looking for

@bucaran is fly checking for custom `node_modules` paths?

**Comment by ghost** [comment]
Just using `require`. We are not behaving any different from other node utilities @Aweary may be already familiar with.

Also, if his global `node_modules` is under `$HOME` I guess he also has `$HOME` in his `$PATH` (a bit unusual but that's up to him) he should be able to run `fly` in the CLI.

**Comment by MadcapJake** [comment]
@Aweary two things come to mind:

1) Maybe check your `.npmrc` or just try using `nvm`'s default config.
2) You also might want to try using babel to compile the flyfile and see what babel says/outputs. The stacktrace makes it seem like maybe babel is silently failing.

**Comment by aweary** [comment]
It's not configuration related, as I've completed removed and reinstalled `node` using both `nvm` and just directly via the install package, and it occurs regardless. Babel works fine if I manually parse the file. 

For the `Flyfile` I posted in the OP, I get the following output directly compiliing with `babel`:

``` js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});
exports['default'] = regeneratorRuntime.mark(function callee$0$0() {
    return regeneratorRuntime.wrap(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return this.source('secret').target('backup');

            case 2:
            case 'end':
                return context$1$0.stop();
        }
    }, callee$0$0, this);
});
module.exports = exports['default'];
```

Nothing really seems out of place, and I can't figure out the root cause, so I'm going to close this issue for now. If I find a root cause or way to reproduce, I'll reopen it.

**Comment by ghost** [comment]
That's the correct output.

@Aweary The problem is not with your node, however. 

Question, why are your global node packages installed to $HOME? I wonder if this could have something to do with this.

**Comment by aweary** [comment]
I set the install location to $HOME a while back while working on some unrelated stuff, debugging and whatnot. But again, reinstalling Node resets all configurations. All my global packages are in the standard location, and the issue still occurs.

**Comment by ghost** [comment]
Gotcha. @Aweary Would you mind uploading the failling project to GitHub so we can all have a look at it?

We are getting close to fix this.

**Comment by aweary** [comment]
Here's a copy of the entire folder, `node_modules` and all: https://github.com/Aweary/flytest

**Comment by ghost** [comment]
Went over it. Removed the Fly link and simply ran specifying the full path to `bin/index.js` and it was also OK. @MadcapJake Do you want to give it a shot?

**Comment by MadcapJake** [comment]
Yeah I was able to get that working.  Tried a local and global install.  You might consider trying another language to see if it produces the same bug as `Unexpected token export` makes it seem like it can't recognize the syntax.  If another compiled language causes a similar unrecognized-syntax error then that might be a lead on solving this.

Also, have you tried running fly with the pre-compiled flyfile? And maybe try some other ES6 examples from the Fly repo.  Did using a local fly and npm scripts solve the issue? At least that would be a temporary solution.

**Comment by iamstarkov** [comment]
Fly needs improvement to show both local and global versions, like gulp do:

```
➜  jsunderhood git:(master) ✗ gulp -v
[14:44:47] Requiring external module babel-core/register
[14:44:47] CLI version 3.9.0
[14:44:47] Local version 3.9.0
```

**Comment by aweary** [comment]
I'm seeing the same error in other `babel` projects, so I've opened https://github.com/babel/babel/issues/2085 to see what they can find out. It just seems that `babel` isn't compiling imports at all for me.

**Comment by aweary** [comment]
So what seems to be happening is that `fly` is not compiling the `Flyfile` before it's required by the `spawn.js` file. If we can go over how `fly` compiles the `Flyfile.babel.js` file then we can probably find out why exactly this is occurring.

**Comment by ghost** [comment]
That is handled in fly-util [here](https://github.com/flyjs/fly-util/blob/master/src/findPath.js).

``` js
const path = yield find(file)
```

Look at the [`hook`](https://github.com/flyjs/fly-util/blob/master/src/findPath.js#L30) function.

Fly only needs to transpile one file, the Flyfile. I looked at the issue you opened in babel, but that still does not explain why it works for everyone else :disappointed_relieved:

**Comment by ghost** [comment]
@Aweary I think I've found the problem, it's a bug in our end :sob: 

Basically:

```
fly -f path/to/Flyfile.js # BAD
```

<hr>

```
fly -f path/to/ # OK
```

In the first case, Fly is not property installing the require hook, hence the problem. I will make sure to fix this by 0.4.0

**Comment by aweary** [comment]
:open_mouth: 

After all that, I was totally convinced I had some crazy configuration issue going on.

At least I know my system isn't just broken!

**Comment by ghost** [comment]
My bad @Aweary  :smile:

**Comment by ghost** [comment]
@Aweary Please upgrade to [@0.4.0](https://github.com/flyjs/fly/releases) and let me know how things go :smile:

**Comment by aweary** [comment]
`0.4.0` has fixed the issue for me! Looks like we're good from here

:+1:

**Comment by ghost** [comment]
Yay! (o^-')b !
